### PR TITLE
 Minimal implementation of onionpeer object

### PR DESCRIPTION
**Pull request by g1itch**
Hello!

In this changes each node having onion hidden service sends an object with objectType 4 and TTL 1 week.

### CODE REVIEW

**Comment by g1itch** [comment]
So it looks like almost all code setting an inventory item is wrong because it should use `buffer()` for payload and tag:
```patch
diff --git a/src/api.py b/src/api.py
index 1e485209..96e31aaf 100644
--- a/src/api.py
+++ b/src/api.py
@@ -1219,7 +1219,7 @@ class MySimpleXMLRPCRequestHandler(SimpleXMLRPCRequestHandler):
         objectType = 2
         TTL = 2.5 * 24 * 60 * 60
         Inventory()[inventoryHash] = (
-            objectType, toStreamNumber, encryptedPayload,
+            objectType, toStreamNumber, buffer(encryptedPayload),
             int(time.time()) + TTL, ''
         )
         with shared.printLock:
@@ -1275,7 +1275,8 @@ class MySimpleXMLRPCRequestHandler(SimpleXMLRPCRequestHandler):
         objectType = 1  # .. todo::: support v4 pubkeys
         TTL = 28 * 24 * 60 * 60
         Inventory()[inventoryHash] = (
-            objectType, pubkeyStreamNumber, payload, int(time.time()) + TTL, ''
+            objectType, pubkeyStreamNumber, buffer(payload),
+            int(time.time()) + TTL, ''
         )
         with shared.printLock:
             print 'broadcasting inv within API command disseminatePubkey with hash:', hexlify(inventoryHash)
diff --git a/src/class_singleWorker.py b/src/class_singleWorker.py
index 5076e0ce..b27cb4c7 100644
--- a/src/class_singleWorker.py
+++ b/src/class_singleWorker.py
@@ -281,7 +281,7 @@ class singleWorker(threading.Thread, StoppableThread):
         inventoryHash = calculateInventoryHash(payload)
         objectType = 1
         Inventory()[inventoryHash] = (
-            objectType, streamNumber, payload, embeddedTime, '')
+            objectType, streamNumber, buffer(payload), embeddedTime, '')
 
         logger.info('broadcasting inv with hash: %s', hexlify(inventoryHash))
 
@@ -363,7 +363,7 @@ class singleWorker(threading.Thread, StoppableThread):
         inventoryHash = calculateInventoryHash(payload)
         objectType = 1
         Inventory()[inventoryHash] = (
-            objectType, streamNumber, payload, embeddedTime, '')
+            objectType, streamNumber, buffer(payload), embeddedTime, '')
 
         logger.info('broadcasting inv with hash: %s', hexlify(inventoryHash))
 
@@ -451,8 +451,8 @@ class singleWorker(threading.Thread, StoppableThread):
         inventoryHash = calculateInventoryHash(payload)
         objectType = 1
         Inventory()[inventoryHash] = (
-            objectType, streamNumber, payload, embeddedTime,
-            doubleHashOfAddressData[32:]
+            objectType, streamNumber, buffer(payload), embeddedTime,
+            buffer(doubleHashOfAddressData[32:])
         )
 
         logger.info('broadcasting inv with hash: %s', hexlify(inventoryHash))
@@ -645,7 +645,8 @@ class singleWorker(threading.Thread, StoppableThread):
             inventoryHash = calculateInventoryHash(payload)
             objectType = 3
             Inventory()[inventoryHash] = (
-                objectType, streamNumber, payload, embeddedTime, tag)
+                objectType, streamNumber, buffer(payload), embeddedTime,
+                buffer(tag))
             logger.info(
                 'sending inv (within sendBroadcast function)'
                 ' for object: %s',
@@ -1240,7 +1241,8 @@ class singleWorker(threading.Thread, StoppableThread):
             inventoryHash = calculateInventoryHash(encryptedPayload)
             objectType = 2
             Inventory()[inventoryHash] = (
-                objectType, toStreamNumber, encryptedPayload, embeddedTime, '')
+                objectType, toStreamNumber, buffer(encryptedPayload),
+                embeddedTime, '')
             if BMConfigParser().has_section(toaddress) or \
                not protocol.checkBitfield(behaviorBitfield, protocol.BITFIELD_DOESACK):
                 queues.UISignalQueue.put((
@@ -1400,7 +1402,7 @@ class singleWorker(threading.Thread, StoppableThread):
         inventoryHash = calculateInventoryHash(payload)
         objectType = 1
         Inventory()[inventoryHash] = (
-            objectType, streamNumber, payload, embeddedTime, '')
+            objectType, streamNumber, buffer(payload), embeddedTime, '')
         logger.info('sending inv (for the getpubkey message)')
         queues.invQueue.put((streamNumber, inventoryHash))
 

```

File: src/class_objectProcessor.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 from binascii import hexlify
 from subprocess import call  # nosec
 
+import knownnodes
```
Comment (omkar1117) [inline src/class_objectProcessor.py]: Pep8 standards please

File: src/class_singleWorker.py [hunk 2]
```diff
@@ -119,13 +119,21 @@ def run(self):
...
         self.stop.wait(10)
 
-        if state.shutdown == 0:
-            # just in case there are any pending tasks for msg
-            # messages that have yet to be sent.
-            queues.workerQueue.put(('sendmessage', ''))
-            # just in case there are any tasks for Broadcasts
-            # that have yet to be sent.
-            queues.workerQueue.put(('sendbroadcast', ''))
+        if state.shutdown > 0:
```
Comment (omkar1117) [inline src/class_singleWorker.py]: can't we write like this
if not state.shutdown:

**Comment by g1itch** [comment]
> Could you please write the import statements first please

If you mean "write local application specific `import ...` statements before standard library `from ... import ...`", the answer is no, because that directly violates pep8 recommendations: https://www.python.org/dev/peps/pep-0008/#imports

File: src/protocol.py [hunk 3]
```diff
@@ -48,6 +48,7 @@
 OBJECT_PUBKEY = 1
 OBJECT_MSG = 2
 OBJECT_BROADCAST = 3
+OBJECT_ONIONPEER = 4
```
Comment (PeterSurda) [inline src/protocol.py]: How about 0x746f72 (hex for tor)?

Comment (g1itch) [reply src/protocol.py]: OK, I changed

**Pull request merged**